File: src/main/java/io/github/shaksternano/mediamanipulator/image/VideoFrame.java
package io.github.shaksternano.mediamanipulator.image;

public sealed interface VideoFrame<T> permits ImageFrame, AudioFrame {

    /**
     * The content of the frame.
     *
     * @return The content of the frame.
     */
    T content();

    /**
     * The duration of the frame in microseconds.
     *
     * @return The duration of the frame in microseconds.
     */
    double duration();

    /**
     * The timestamp of the frame in microseconds.
     *
     * @return The timestamp of the frame in microseconds.
     */
    long timestamp();
}

File: core/src/main/kotlin/media/VideoFrame.kt
package com.shakster.borgar.core.media

import org.bytedeco.javacv.Frame
import java.awt.image.BufferedImage
import kotlin.time.Duration

data class VideoFrame<T>(
    val content: T,
    val duration: Duration,
    val timestamp: Duration,
)

typealias ImageFrame = VideoFrame<BufferedImage>
typealias AudioFrame = VideoFrame<Frame>
